wdl-grammar/src/tree/dive.rs

```diff
@@ -0,0 +1,265 @@
+    _phantom: PhantomData<L>,
+}
+
+impl<L, M, I> DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    /// Creates a new [`DiveIterator`].
+    ///
+    /// Note that this iterator traverses elements as a breadth-first traversal
+    /// (otherwise known as level-order).
+    pub fn new(root: rowan::SyntaxElement<L>, match_predicate: M, ignore_predicate: I) -> Self {
+        let mut nodes = VecDeque::new();
+        nodes.push_back(root);
+
+        Self {
+            nodes,
+            match_predicate,
+            ignore_predicate,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<L, M, I> Iterator for DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    type Item = rowan::SyntaxElement<L>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(element) = self.nodes.pop_front() {
+            match &element {
+                rowan::SyntaxElement::Node(node) => {
+                    if !(self.ignore_predicate)(node) {
+                        self.nodes.extend(node.children_with_tokens());
+                    }
+                }
+                // NOTE: tokens have no children to explore.
+                rowan::SyntaxElement::Token(_) => {}
+            }
+
+            if (self.match_predicate)(&element) {
+                // NOTE: this is an inexpensive clone of a red node.
+                return Some(element.clone());
+            }
+        }
+
+        None
+    }
+}
+
+impl<L, M, I> FusedIterator for DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+}
+
+/// Elements of a syntax tree upon which a dive can be performed.
+pub trait Divable<L>
+where
+    L: Language,
+{
+    /// Visits every element in a syntax tree and accumulates all
+    /// [`SyntaxElement`](rowan::SyntaxElement)s for which the `match_predicate`
+    /// evaluates to `true`.
+    ///
+    /// No guarantees are made about the order in which elements will be
+    /// traversed or collected.
+    fn dive<M>(&self, match_predicate: M) -> impl Iterator<Item = rowan::SyntaxElement<L>>
+    where
+        M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    {
+        self.dive_with_ignore(match_predicate, |_| false)
+    }
+
+    /// Visits every element in a syntax tree and accumulates all
+    /// [`SyntaxElement`](rowan::SyntaxElement)s for which the `match_predicate`
+    /// evaluates to `true`. No children nodes will be searched for
+    /// [`SyntaxNode`](rowan::SyntaxNode)s that are visited and for which the
+    /// `ignore_predicate` evaluates to `true`. In this way, you can ignore
+    /// entire sections of the tree.
+    ///
+    /// No guarantees are made about the order in which elements will be
+    /// traversed or collected.
```

Comment:
Can you clarify what you mean here? You're doing a breadth-first traversal. Do you just mean that the `ignore` component breaks the order guarantee? It should still be breadth-first-ish right?

Reply:
No, it's not that the `ignore` component breaks the guarantee but rather that, by virtue of the way the elements are accumulated and stored, the traversal is different. I would consider this to be fully breadth-first, as every element is inspect in a BFS order (though not every element is returned).